pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -168,31 +417,119 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
 		// Make a deep copy so we don't mutate the shared cache
 		copy, err := scheme.Scheme.DeepCopy(replicas[i])
 		if err != nil {
-			return err
+			return &status, err
 		}
 		replica := copy.(*v1.Pod)
-		if err := ssc.podControl.UpdateStatefulPod(set, replica); err != nil {
-			return err
+		if err := ssc.podControl.UpdateStatefulPod(updateSet, replica); err != nil {
+			return &status, err
 		}
 	}
 
 	// At this point, all of the current Replicas are Running and Ready, we can consider termination.
 	// We will wait for all predecessors to be Running and Ready prior to attempting a deletion.
 	// We will terminate Pods in a monotonically decreasing order over [len(pods),set.Spec.Replicas).
-	// Note that we do not resurrect Pods in this interval.
-	if unhealthy > 0 && monotonic {
-		glog.V(2).Infof("StatefulSet %s is waiting on %d Pods", set.Name, unhealthy)
-		return nil
-	}
+	// Note that we do not resurrect Pods in this interval. Also not that scaling will take precedence over
+	// updates.
 	for target := len(condemned) - 1; target >= 0; target-- {
-		glog.V(2).Infof("StatefulSet %s terminating Pod %s", set.Name, condemned[target])
+		// wait for terminating pods to expire
+		if isTerminating(condemned[target]) {
+			glog.V(4).Infof(
+				"StatefulSet %s/%s is waiting for Pod %s to Terminate prior to scale down",
+				set.Namespace,
+				set.Name,
+				condemned[target].Name)
+			// block if we are in monotonic mode
+			if monotonic {
+				return &status, nil
+			}
+			continue
+		}
+		// if we are in monotonic mode and the condemned target is not the first unhealthy Pod block
+		if !isRunningAndReady(condemned[target]) && monotonic && condemned[target] != firstUnhealthyPod {
```

Comment:
is this checking saying that before we can delete or scale down, the extra pods must be healthy and running ? Is it for safety of clustered applications ?